File: examples/client-gen/tictactoe/accounts/game.py
import typing
from dataclasses import dataclass
from solders.pubkey import Pubkey
from solana.rpc.async_api import AsyncClient
from solana.rpc.commitment import Commitment
import borsh_construct as borsh
from anchorpy.coder.accounts import ACCOUNT_DISCRIMINATOR_SIZE
from anchorpy.error import AccountInvalidDiscriminator
from anchorpy.utils.rpc import get_multiple_accounts
from anchorpy.borsh_extension import BorshPubkey
from ..program_id import PROGRAM_ID
from .. import types


class GameJSON(typing.TypedDict):
    players: list[str]
    turn: int
    board: list[list[typing.Optional[types.sign.SignJSON]]]
    state: types.game_state.GameStateJSON


@dataclass
class Game:
    discriminator: typing.ClassVar = b"\x1bZ\xa6}Jdy\x12"
    layout: typing.ClassVar = borsh.CStruct(
        "players" / BorshPubkey[2],
        "turn" / borsh.U8,
        "board" / borsh.Option(types.sign.layout)[3][3],
        "state" / types.game_state.layout,
    )
    players: list[Pubkey]
    turn: int
    board: list[list[typing.Optional[types.sign.SignKind]]]
    state: types.game_state.GameStateKind

    @classmethod
    async def fetch(
        cls,
        conn: AsyncClient,
        address: Pubkey,
        commitment: typing.Optional[Commitment] = None,
        program_id: Pubkey = PROGRAM_ID,
    ) -> typing.Optional["Game"]:
        resp = await conn.get_account_info(address, commitment=commitment)
        info = resp.value
        if info is None:
            return None
        if info.owner != program_id:
            raise ValueError("Account does not belong to this program")
        bytes_data = info.data
        return cls.decode(bytes_data)

    @classmethod
    async def fetch_multiple(
        cls,
        conn: AsyncClient,
        addresses: list[Pubkey],
        commitment: typing.Optional[Commitment] = None,
        program_id: Pubkey = PROGRAM_ID,
    ) -> typing.List[typing.Optional["Game"]]:
        infos = await get_multiple_accounts(conn, addresses, commitment=commitment)
        res: typing.List[typing.Optional["Game"]] = []
        for info in infos:
            if info is None:
                res.append(None)
                continue
            if info.account.owner != program_id:
                raise ValueError("Account does not belong to this program")
            res.append(cls.decode(info.account.data))
        return res

    @classmethod
    def decode(cls, data: bytes) -> "Game":
        if data[:ACCOUNT_DISCRIMINATOR_SIZE] != cls.discriminator:
            raise AccountInvalidDiscriminator(
                "The discriminator for this account is invalid"
            )
        dec = Game.layout.parse(data[ACCOUNT_DISCRIMINATOR_SIZE:])
        return cls(
            players=dec.players,
            turn=dec.turn,
            board=list(
                map(
                    lambda item: list(
                        map(
                            lambda item: (
                                None if item is None else types.sign.from_decoded(item)
                            ),
                            item,
                        )
                    ),
                    dec.board,
                )
            ),
            state=types.game_state.from_decoded(dec.state),
        )

    def to_json(self) -> GameJSON:
        return {
            "players": list(map(lambda item: str(item), self.players)),
            "turn": self.turn,
            "board": list(
                map(
                    lambda item: list(
                        map(
                            lambda item: (None if item is None else item.to_json()),
                            item,
                        )
                    ),
                    self.board,
                )
            ),
            "state": self.state.to_json(),
        }

    @classmethod
    def from_json(cls, obj: GameJSON) -> "Game":
        return cls(
            players=list(map(lambda item: Pubkey.from_string(item), obj["players"])),
            turn=obj["turn"],
            board=list(
                map(
                    lambda item: list(
                        map(
                            lambda item: (
                                None if item is None else types.sign.from_json(item)
                            ),
                            item,
                        )
                    ),
                    obj["board"],
                )
            ),
            state=types.game_state.from_json(obj["state"]),
        )

File: examples/tic-tac-toe/generated-client/accounts/Game.ts
import { PublicKey, Connection } from "@solana/web3.js"
import BN from "bn.js" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as borsh from "@project-serum/borsh" // eslint-disable-line @typescript-eslint/no-unused-vars
import * as types from "../types" // eslint-disable-line @typescript-eslint/no-unused-vars
import { PROGRAM_ID } from "../programId"

export interface GameFields {
  players: Array<PublicKey>
  turn: number
  board: Array<Array<types.SignKind | null>>
  state: types.GameStateKind
}

export interface GameJSON {
  players: Array<string>
  turn: number
  board: Array<Array<types.SignJSON | null>>
  state: types.GameStateJSON
}

export class Game {
  readonly players: Array<PublicKey>
  readonly turn: number
  readonly board: Array<Array<types.SignKind | null>>
  readonly state: types.GameStateKind

  static readonly discriminator = Buffer.from([
    27, 90, 166, 125, 74, 100, 121, 18,
  ])

  static readonly layout = borsh.struct([
    borsh.array(borsh.publicKey(), 2, "players"),
    borsh.u8("turn"),
    borsh.array(borsh.array(borsh.option(types.Sign.layout()), 3), 3, "board"),
    types.GameState.layout("state"),
  ])

  constructor(fields: GameFields) {
    this.players = fields.players
    this.turn = fields.turn
    this.board = fields.board
    this.state = fields.state
  }

  static async fetch(c: Connection, address: PublicKey): Promise<Game | null> {
    const info = await c.getAccountInfo(address)

    if (info === null) {
      return null
    }
    if (!info.owner.equals(PROGRAM_ID)) {
      throw new Error("account doesn't belong to this program")
    }

    return this.decode(info.data)
  }

  static async fetchMultiple(
    c: Connection,
    addresses: PublicKey[]
  ): Promise<Array<Game | null>> {
    const infos = await c.getMultipleAccountsInfo(addresses)

    return infos.map((info) => {
      if (info === null) {
        return null
      }
      if (!info.owner.equals(PROGRAM_ID)) {
        throw new Error("account doesn't belong to this program")
      }

      return this.decode(info.data)
    })
  }

  static decode(data: Buffer): Game {
    if (!data.slice(0, 8).equals(Game.discriminator)) {
      throw new Error("invalid account discriminator")
    }

    const dec = Game.layout.decode(data.slice(8))

    return new Game({
      players: dec.players,
      turn: dec.turn,
      board: dec.board.map((item) =>
        item.map((item) => (item && types.Sign.fromDecoded(item)) || null)
      ),
      state: types.GameState.fromDecoded(dec.state),
    })
  }

  toJSON(): GameJSON {
    return {
      players: this.players.map((item) => item.toString()),
      turn: this.turn,
      board: this.board.map((item) =>
        item.map((item) => (item && item.toJSON()) || null)
      ),
      state: this.state.toJSON(),
    }
  }

  static fromJSON(obj: GameJSON): Game {
    return new Game({
      players: obj.players.map((item) => new PublicKey(item)),
      turn: obj.turn,
      board: obj.board.map((item) =>
        item.map((item) => (item && types.Sign.fromJSON(item)) || null)
      ),
      state: types.GameState.fromJSON(obj.state),
    })
  }
}
